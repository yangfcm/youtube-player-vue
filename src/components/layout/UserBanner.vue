<template>
  <div class="app-banner-container">
    <div class="app-banner-image">
      <i class="user secret icon huge" v-if="signedIn === false"></i>
      <img
        v-if="signedIn === true"
        class="ui tiny image circular"
        :src="auth.user && auth.user.picture"
        :alt="auth.user.name"
      />
    </div>

    <div class="app-banner-title">
      <h2 class="ui header app-banner-title-text">
        <span style="margin-right: 3px;"
          >Welcome
          {{ signedIn === true ? ` back, ${auth.user.name}` : "" }}
        </span>
      </h2>
    </div>
  </div>
</template>

<script>
import { mapState, mapGetters } from "vuex";
export default {
  computed: {
    ...mapState(["auth"]),
    ...mapGetters(["signedIn"]),
  },
};
</script>
<style scoped>
.app-banner-container {
  margin-top: 1.5rem;
  margin-bottom: 1.5rem;
  display: flex;
}
.app-banner-image {
  margin-right: 1rem;
  min-height: 80px;
  display: flex;
  align-items: center;
}
.app-banner-title {
  display: flex !important;
  align-items: center;
}
.app-banner-title-text {
  display: flex;
}
@media (max-width: 576px) {
  .app-banner-title-text {
    flex-direction: column;
  }
}
</style>
