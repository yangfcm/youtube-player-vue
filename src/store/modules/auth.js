import axios from "../../settings";
import { SIGN_IN, SIGN_OUT, CATCH_AUTH_ERROR } from "../types";

const state = {
  user: null,
};

const getters = {
  signedIn: (state) => {
    const { user } = state;
    if (user === null) {
      return null;
    }
    if (user) {
      if (Object.keys(user).length === 0) {
        return false;
      } else {
        return true;
      }
    }
  },
  user: (state) => state.user,
};

const mutations = {
  SIGN_IN: (state, payload) => {
    state.user = payload;
  },
  SIGN_OUT: (state) => {
    state.user = {};
  },
  CATCH_AUTH_ERROR: (state) => {
    state.user = {};
  },
};

const actions = {
  signIn: async (context, googleUser) => {
    try {
      const accessToken = `${googleUser.wc.token_type} ${googleUser.wc.access_token}`;
      localStorage.setItem("access_token", accessToken);
      const response = await axios.get(
        "https://www.googleapis.com/oauth2/v1/userinfo?alt=json",
        {
          headers: {
            Authorization: accessToken,
          },
        }
      );
      context.commit(SIGN_IN, response.data);
    } catch (err) {
      context.commit(CATCH_AUTH_ERROR);
    }
  },
  signOut: (context) => {
    localStorage.removeItem("access_token");
    context.commit(SIGN_OUT);
  },
};

export default {
  state,
  getters,
  mutations,
  actions,
};
