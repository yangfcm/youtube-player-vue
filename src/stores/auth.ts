import { ref, computed } from 'vue'
import { defineStore } from 'pinia'
import { type AxiosResponse } from 'axios'
import { AsyncStatus } from '@/settings/types'
import { gAuthAxios } from '@/settings/api'
import type { PlayListsRespone, SubscriptionsResponse } from './types'
import { fetchSubscriptionsAPI, fetchPlayListsAPI } from './api'

type UserInfoResponse = {
  email: string
  email_verified: boolean
  family_name: string
  given_name: string
  locale: string
  name: string
  picture: string
  sub: string
}

type Auth = {
  user: UserInfoResponse | null
  status: AsyncStatus
  subscriptions: {
    status: AsyncStatus
    error: string
    data?: SubscriptionsResponse
    subscriptionIds: Record<string, string>
  }
  playlists: {
    status: AsyncStatus
    error: string
    data?: PlayListsRespone
  }
}

export const useAuthStore = defineStore('auth', () => {
  const auth = ref<Auth>({
    user: null,
    status: AsyncStatus.IDLE,
    subscriptions: {
      status: AsyncStatus.IDLE,
      error: '',
      subscriptionIds: {},
    },
    playlists: {
      status: AsyncStatus.IDLE,
      error: '',
    },
  })

  // function signin(authData: Auth) {
  //   auth.value = authData
  // }

  function signout() {
    auth.value.user = null
    auth.value.status = AsyncStatus.IDLE
    auth.value.subscriptions = {
      status: AsyncStatus.IDLE,
      error: '',
      subscriptionIds: {},
    }
    auth.value.playlists = {
      status: AsyncStatus.IDLE,
      error: '',
    }
  }

  const status = computed(() => {
    return auth.value.status
  })

  const user = computed(() => {
    return auth.value.user
  })

  const isSignedIn = computed(() => {
    if (!auth.value.user) return false
    return true
  })

  const subscriptions = computed(() => {
    return auth.value.subscriptions
  })

  const playlists = computed(() => {
    return auth.value.playlists
  })

  const fetchUserByToken = async (token: string) => {
    try {
      auth.value.status = AsyncStatus.LOADING
      const response: AxiosResponse<UserInfoResponse> = await gAuthAxios.get('/', {
        params: {
          access_token: token,
        },
      })
      auth.value.user = response.data
      auth.value.status = AsyncStatus.SUCCESS
    } catch {
      signout()
    }
  }

  const fetchSubscriptions = async (pageToken?: string) => {
    try {
      auth.value.subscriptions.status = AsyncStatus.LOADING
      auth.value.subscriptions.error = ''
      const options: Record<string, string> = {}
      if (pageToken) options.pageToken = pageToken
      const response = await fetchSubscriptionsAPI(options)
      const currentItems = auth.value.subscriptions.data?.items || []
      auth.value.subscriptions.status = AsyncStatus.SUCCESS
      auth.value.subscriptions.data = {
        ...response.data,
        items: [...currentItems, ...response.data.items],
      }
      response.data.items.forEach(
        (item) =>
          (auth.value.subscriptions.subscriptionIds[item.snippet.resourceId.channelId] = item.id),
      )
    } catch (err: any) {
      auth.value.subscriptions.status = AsyncStatus.FAIL
      auth.value.subscriptions.error = err.message
    }
  }

  const fetchPlayLists = async (pageToken?: string) => {
    try {
      auth.value.playlists.status = AsyncStatus.LOADING
      auth.value.playlists.error = ''
      const options: Record<string, string> = {}
      if (pageToken) options.pageToken = pageToken
      const response = await fetchPlayListsAPI(options)
      const currentItems = auth.value.playlists.data?.items || []
      auth.value.playlists.status = AsyncStatus.SUCCESS
      auth.value.playlists.data = {
        ...response.data,
        items: [...currentItems, ...response.data.items],
      }
    } catch (err: any) {
      auth.value.playlists.status = AsyncStatus.FAIL
      auth.value.playlists.error = err.message
    }
  }

  return {
    status,
    user,
    isSignedIn,
    subscriptions,
    playlists,
    // signin,
    signout,
    fetchUserByToken,
    fetchSubscriptions,
    fetchPlayLists,
    authState: auth,
  }
})
