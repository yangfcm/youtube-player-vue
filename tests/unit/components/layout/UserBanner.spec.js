import { shallowMount } from "@vue/test-utils";
import UserBanner from "@/components/layout/UserBanner.vue";
import { userResponse as user } from "../../store/fixtures/auth";

test("Userbanner component can render correctly when user not logged in", () => {
  const $store = {
    state: {
      auth: {
        user: {},
      },
    },
    getters: {
      signedIn: false,
    },
  };
  const wrapper = shallowMount(UserBanner, {
    mocks: {
      $store,
    },
  });
  expect(wrapper).toMatchSnapshot();
  expect(wrapper.find("i.user.secret.icon").exists()).toBe(true);
  expect(wrapper.find("img").exists()).toBe(false);
});

test("Userbanner component can render correctly when user logged in", () => {
  const $store = {
    state: {
      auth: {
        user,
      },
    },
    getters: {
      signedIn: true,
    },
  };
  const wrapper = shallowMount(UserBanner, {
    mocks: {
      $store,
    },
  });
  expect(wrapper).toMatchSnapshot();
  expect(wrapper.find("i.user.secret.icon").exists()).toBe(false);
  expect(wrapper.find("img").exists()).toBe(true);
  expect(wrapper.text()).toContain(wrapper.vm.$store.state.auth.user.name);
});
