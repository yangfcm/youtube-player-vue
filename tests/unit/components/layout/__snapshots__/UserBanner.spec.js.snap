// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Userbanner component can render correctly when user logged in 1`] = `
<div class="app-banner-container">
  <div class="app-banner-image">
    <!----> <img src="https://lh5.googleusercontent.com/-MuN5BmJ3lFM/AAAAAAAAAAI/AAAAAAAAAAA/AMZuucmzrAOkewOEoGUPasQOSYSk3Gl9nA/photo.jpg" alt="sam test" class="ui tiny image circular"></div>
  <div class="app-banner-title">
    <h2 class="ui header app-banner-title-text"><span style="margin-right: 3px;">Welcome
         back, sam test
      </span></h2>
  </div>
</div>
`;

exports[`Userbanner component can render correctly when user not logged in 1`] = `
<div class="app-banner-container">
  <div class="app-banner-image"><i class="user secret icon huge"></i>
    <!---->
  </div>
  <div class="app-banner-title">
    <h2 class="ui header app-banner-title-text"><span style="margin-right: 3px;">Welcome
        
      </span></h2>
  </div>
</div>
`;
